lib/linked_list.rb

```diff
@@ -18,71 +20,185 @@ def initialize
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_max
-      raise NotImplementedError
+      current_node = @head
+      max_value = current_node.data unless current_node.nil?
+
+      until current_node.nil?
+        max_value = current_node.data if current_node.data > max_value
+
+        current_node = current_node.next
+      end
+
+      return max_value
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_min
-      raise NotImplementedError
+      current_node = @head
+      min_value = current_node.data unless current_node.nil?
+
+      until current_node.nil?
+        min_value = current_node.data if current_node.data < min_value
+
+        current_node = current_node.next
+      end
+
+      return min_value
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      current_node = @head
+      length = 0
+
+      until current_node.nil?
+        length += 1
+
+        current_node = current_node.next
+      end
+
+      return length
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      length = self.length
+      return nil if length <= index
+      return @head.data if index == 0
+
+      current_index = 0
+      current_node = @head
+
+      until current_index == index
+        current_index += 1
+        current_node = current_node.next
+      end
+
+      return current_node.data
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit
-      raise NotImplementedError
+      return nil if @head.nil?
+      current_node = @head
+
+      until current_node.nil?
+        puts current_node.data
+
+        current_node = current_node.next
+      end
+
+      return
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(value)
-      raise NotImplementedError
+      return if @head.nil?
+
+      if @head.data == value
+        @head = nil if @head.next.nil?
+        @head = @head.next unless @head.nil?
+        return
+      end

```

Comment:
This can be simpler, but since you have a doubly linked list you also need to set the previous.

```suggestion
      if @head.data == value
        @head = @head.next
        @head.previous = nil
        return
      end
```